### ISSUE TITLE
Please, fix IUCN Red List API

### ISSUE THREAD

**Issue by Archilegt** [NaturalHistoryMuseum/scratchpads2#6564]
**Describe the bug**
The UICN API is not working on taxon description pages

**To Reproduce**
See for example:
_Scolopendra abnormis_ https://myriatrix.myspecies.info/taxonomy/term/9019/descriptions
which corresponds to _Scolopendra abnormis_ https://www.iucnredlist.org/species/20042/9138315
This API was working recently as should not be difficult to fix. Aggregating species conservation information was one of the key outputs of Scratchpads and still sets it apart and ahead from other platforms.

**Comment by therobyouknow** [comment]
confirming I can see it is not working in that "No results.", "Unable to fetch data from NCBI." and "Unable to fetch data from IUCN." also as shown in screenshot below.

![image](https://user-images.githubusercontent.com/1080646/178941303-f7ca8a95-dc0b-4260-a80f-081c4b74ba06.png)

Do you know when it was last working, @Archilegt ?

I amd starting to analysing possible root causes of the problem this morning.

**Comment by therobyouknow** [comment]
turned on logging (dblog / watchdog) to get further error information. Would want to turn this off - back to original state after analysis.

**Comment by therobyouknow** [comment]
nothing stands out in the dblogs at the moment, see below, so I'll switch them off.


![image](https://user-images.githubusercontent.com/1080646/178958113-360ffcaf-0aca-4ce6-a97f-12dd10a6fd15.png)

Next steps: I'll make a copy of the site on my own machine for further analysis using xdebug.

I started looking at the custom scratchpads code, there is code here: sites/all/modules/custom/scratchpads/scratchpads_species/modules/ncbi/ncbi.module that makes the call the NCBI.

As expected and correct: there are NCBI URLs in module - API endpoints to fetch the NCBI data.
 
A quick test with one of them and some made-up id value show it to work https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=443 So for this API call at least, it is still available from the provider.

Other things to consider would be firewall between the scratchpads site and the outside world - whether that needs changing.

As I've said I'll set this up on my own machine. This will provide 2 things - 1 - to see if the API calls work on my own machine - which doesn't have the production server firewall. And if so, then we know it could be a firewall issue or something particular about production. Secondly - 2 - running the copy of the site on my local machine will afford me the best tools for debugging - xdebug, to step through the code to see why these API calls are not working - if the outcome of 1 is the same as production and firewall is not an issue.

**Comment by therobyouknow** [comment]
Token requested for using https://apiv3.iucnredlist.org/ 

Will await response from https://www.iucnredlist.org/ team.

**Comment by Archilegt** [comment]
Can confirm that Scratchpads needs the token.
Feel free to ping me if a working session is needed for mapping compatibility of previous UICN API and SP code to current UICN API and new SP code. I evaluated a species for the Red List of Peru not so long ago and data structure is still fresh in my mind.
Issue can be renamed as "Upgrade UICN API" and migrated to Scratchpads evolution.

**Comment by therobyouknow** [comment]
**making use of new v3 API to implement solution, potential refactoring code**

Debugging IUCN response for initial try at calling v3 api, using previous v2 calls.  Would be nice if IUCN could provide a v2 vs v3 document to help users of v2 move over to v3, e.g. what new calls/endpoints in v3 can be used in place of v2 counterparts.

<img width="539" alt="image" src="https://user-images.githubusercontent.com/1080646/203730365-aa75d1a0-9d9a-429d-a2c2-134f5b171d51.png">

**Comment by Archilegt** [comment]
Test calling `Scolopendra abnormis`. Post result.
Also, I agree. I would myself need to see the documentation of the previous API to do the mapping to the [documentation of the new API](https://apiv3.iucnredlist.org/api/v3/docs).

For Scratchpads evolution, we contemplated the creation of a "Country Content Type" under which we would display taxonomic lists (mainly species but also higher categories) per country. Note the synergy with the UICN API v3 output:

```
Species by country
To get a list of species by country isocode:
/api/v3/country/getspecies/:country?token='YOUR TOKEN'
Replace :country with the ISO-code of the country for which you want the list of species
Example: All species for Azerbaijan (AZ)
```

**Comment by therobyouknow** [comment]
What would you like to see displayed from IUCN on your Scratchpads site, @Archilegt ?

**Comment by therobyouknow** [comment]
**implementing some iucn info display, using existing open source tools, contacted IUCN Red List in Cambridge UK**

This looks very good because it looks like it covers at least some of what we used to do: https://github.com/jbroutier/iucn-api-client

I say "what we used to do" according to the training documentation here for using IUCN with Scratchpads: https://scratchpads.readthedocs.io/en/latest/classification/integrating-external-data.html

Screenshot below taken from that documentation.

![image](https://user-images.githubusercontent.com/1080646/203787648-ff788fa0-4481-43d8-9d97-9ea53db24464.png)

As you would see from that above link, our training documentation appears at the moment to be very brief about IUCN, what it is and why we chose what we wanted to display and how this would be useful to scratchpads users.

That, said, the screenshot does show a hierarchical taxonomy and looking at that library, https://github.com/jbroutier/iucn-api-client, I can see, from their documentation, that it appears to be straightforward to display the hierarchical taxonomy in terms of development steps:

> try {
>     $species = $client->getSpeciesByName('Ailurus fulgens');
> } catch (\IucnApi\Exception\IucnApiException $exception) {
>     // Something doesn't look good…
> }
> 
> echo $species->getKigdom();         // ANIMALIA
> echo $species->getPhylum();         // CHORDATA
> echo $species->getClass();          // MAMMALIA
> echo $species->getOrder();          // CARNIVORA
> echo $species->getFamily();         // AILURIDAE
> echo $species->getGenus();          // Ailurus
> echo $species->getMainCommonName(); // Red Panda


Next steps: I will try to set this up locally.



I have contacted IUCN Red List by telephone can got a response from someone who works nearby the team in their building and I have sent an email to the IUCN Red List support email.  While I wait for their response to my query about a document showing how to migrate from v2 to v3, I will continue my own research, and update them with significant findings.

**Comment by therobyouknow** [comment]
**implementation in progress, troubleshooting bugs in code to call IUCN**

I now have https://github.com/jbroutier/iucn-api-client installed into Scratchpads codebase and am now working on overcoming some bugs with the code.

https://github.com/jbroutier/iucn-api-client/issues/1

Hopefully this won't be too involved; re-using another project ought to have benefits instead of writing from scratch. But if the problems are numerous and/or involved then I may have to revisit that decision.

**Comment by therobyouknow** [comment]
**initially working IUCN data coming back**

<img width="771" alt="image" src="https://user-images.githubusercontent.com/1080646/204180904-6a47ccf3-853a-4f56-b0d9-60cda2d1bd07.png">

This is as a result of example new v3 API call:

> https://apiv3.iucnredlist.org/api/v3/species/Scolopendra%20abnormis?token=<kept secret>

which returns:

`{"name":"Scolopendra abnormis","result":[{"taxonid":20042,"scientific_name":"Scolopendra abnormis","kingdom":"ANIMALIA","phylum":"ARTHROPODA","class":"CHILOPODA","order":"SCOLOPENDROMORPHA","family":"SCOLOPENDRIDAE","genus":"Scolopendra","main_common_name":"Serpent Island Centipede","authority":"J.G.E. Lewis &amp; P. Daszak, 1996","published_year":1996,"assessment_date":"1996-08-01","category":"VU","criteria":"D2","population_trend":null,"marine_system":false,"freshwater_system":false,"terrestrial_system":true,"assessor":"Pearce-Kelly, P.","reviewer":"","aoo_km2":null,"eoo_km2":null,"elevation_upper":null,"elevation_lower":null,"depth_upper":null,"depth_lower":null,"errata_flag":null,"errata_reason":null,"amended_flag":null,"amended_reason":null}]}
`

I have modified the copy of https://github.com/jbroutier/iucn-api-client that I put into Scratchpads code base:
- made the project compatible with php 7.2. Well, _so far_ as in the species data parsing works as per screenshot. There may be further similar work to do if we want to use other not yet used functionality of the project
- crucially, adjusted the project to provide the base uri and query for the unique token api key passed in to `Client::getSpeciesByName`:

`            $response = $this->httpClient->request('GET', $url, ['base_uri' => variable_get('iucn_api_v3_url', FALSE), 'query' => ['token' => variable_get('iucn_api_v3_api_key_token', FALSE)]]);`

And so far so good...

But we really need to clarify and make the requirements complete.  All I have to go on is:
- the fact that IUCN doesn't work but do we have screenshots of what the page looks like when it did work?
- the https://scratchpads.readthedocs.io/en/latest/classification/integrating-external-data.html section on IUCN with the screenshot.

That's not enough in terms of requirements.

**Comment by therobyouknow** [comment]
Final dev testing, pre-staging hindered by this seen in debugging. Will reload database on local setup.

<img width="959" alt="image" src="https://user-images.githubusercontent.com/1080646/205119166-7fa96e99-fb59-46e1-909a-f6a1e3e8737f.png">

**Comment by therobyouknow** [comment]
Looking better now. Issue was how to iterate over php object. Fixed:

<img width="937" alt="image" src="https://user-images.githubusercontent.com/1080646/205277232-cb279841-c914-4fbb-a119-f7ab59123988.png">

**Comment by Archilegt** [comment]
A partial view of previous styling can be found here:
https://training2.readthedocs.io/en/latest/classification/integrating-external-data.html

![image](https://user-images.githubusercontent.com/60008120/205279221-36efcc15-51b5-4eaf-bf6b-7a0b39d832a7.png)

Something very important to see in that screenshot is that critical information such as source of the data, terms of use, attribution, and an outlink are displayed on top.
Note also how the data is aggregated into two second-level headers, the first being Taxonomy and the second Assessment Information.

**Comment by therobyouknow** [comment]
@Archilegt yes, thank you, I am aware of this previously: https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6564#issuecomment-1326412851

This isn't a complete requirement I'm afraid. I would suggest I release an initial fixed IUCN data retrieval and we refine this afterwards once we have a full set of requirements.

**Comment by therobyouknow** [comment]
This is what the full output looks like at the moment. Again, as I say, we can refine for next release. I want to make the releases more frequent. I would not want to hold up the release that has the IUCN work so far. Something is better than nothing.

<img width="540" alt="image" src="https://user-images.githubusercontent.com/1080646/205280667-d65c9c05-7050-4e08-9075-224556a00645.png">

**Comment by therobyouknow** [comment]
Moreover, this issue, ticket title "Please, fix UICN API" has been fulfilled by this solution, because we are now using the current v3 IUCN redlist API and getting data back.  Anything else is outside the scope of this issue and should be in another ticket.

With more frequent releases, we can have a more continuous improvement. Thanks for feeding back.

**Comment by therobyouknow** [comment]
**testing in prep for release**

working on branch `rob/release-sprint-2022-11-14-to-2022-11-21`  - test passed.

**Comment by therobyouknow** [comment]
*testing in prep for release**

now testing on another dev machine, to make sure that there is not the table-flip moment of "works for me".

related: https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6611

**Comment by therobyouknow** [comment]
**testing in prep for release**

Branch `rob/release-sprint-2022-11-14-to-2022-11-21` working on other machine.  I did this because of the [git version control work with 3rd party libraries](https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6611), to make sure all files present.  Otherwise, most of the time I would not want to check on 2 different machines or separate copies of the git repo.

![image](https://user-images.githubusercontent.com/1080646/205304261-235b9806-b426-497b-99a4-13aaefdf9c97.png)

Checked `sites/all/libraries/vendor/composer/platform_check.php` is now version controlled - check for php version disabled because we are using an older version, 7.2.

At this point, I would now proceed to testing the release branch `rob/release-sprint-2022-11-14-to-2022-11-21` on a staging site in production, using aegir ci/cd tool to deploy.

**Comment by therobyouknow** [comment]
[New plan](https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6564#issuecomment-1359272299) working so far. API call successful. Now need to output structure in a nice format.

<img width="960" alt="image" src="https://user-images.githubusercontent.com/1080646/208729029-d596949d-41a5-4b3c-80b1-8bb54d7cadfd.png">

**Comment by therobyouknow** [comment]
Working iucn display using new plan

<img width="482" alt="image" src="https://user-images.githubusercontent.com/1080646/208980916-25b09b70-5923-41d2-8b51-36059d415cf7.png">

**Comment by therobyouknow** [comment]
Re-testing iucn on local.  looks like IUCN API down.

Our side seems to be fine - as xdebug shows:

<img width="540" alt="image" src="https://user-images.githubusercontent.com/1080646/211346107-feb73a2c-4a9d-4284-9870-2125e75d4a2e.png">

API landing page down:

<img width="812" alt="image" src="https://user-images.githubusercontent.com/1080646/211346182-e231fd32-b95f-4a00-8683-cedb60ddf709.png">


https://apiv3.iucnredlist.org/api/v3/species/Scolopendra%20abnormis?token=37cde8e5ca269be11d193497becc83c9ad55cb6399c16dadbe1faae61448e199

also bad gateway.

**Comment by therobyouknow** [comment]
Fix now released (in 2.11.1) to your site @Archilegt 

https://myriatrix.myspecies.info/taxonomy/term/9019/descriptions

![image](https://user-images.githubusercontent.com/1080646/212710848-d8120179-4d22-4f01-afaf-868f5ae9e9b1.png)

**Comment by Archilegt** [comment]
Thank you, Rob. I see it working on my side. We will have to meet to decide how to deal with fields that we don't need and how to retrieve fields that are desirable and that should be present at the source. I haven't checked the new UICN thoroughly, but the previous one was serving text information that I don't see now in Scratchpads. We should also discuss about crediting the source. Do you think that it is better to keep this issue open or would you like to have a new one on improving the UICN API integration? Whatever is best for you.
